Should be able to restore even when axes are renamed

Such as restoring from: https://github.com/prjemian/hklpy2/blob/464688282128a5c500a0ff47da0785d1d9917f52/hklpy2/tests/fourc-configuration.yml#L25-L32

Now, it raises a KeyError:

```py
_____________________________________ test_full_position[bissector-keys1-context1-None-fourc-configuration.yml] ______________________________________

mode = 'bissector', keys = ['h', 'k', 'l', 'omega', 'chi', 'phi', ...], context = <contextlib.nullcontext object at 0x7fac11c69950>, expected = None
config_file = 'fourc-configuration.yml'

    @pytest.mark.parametrize(
        "mode, keys, context, expected, config_file",
        [
            ["bissector", "h k l omega chi phi tth".split(), does_not_raise(), None, "e4cv_orient.yml"],
            # FIXME: next test (has renamed axes) should pass with does_not_raise()
            ["bissector", "h k l omega chi phi tth".split(), does_not_raise(), None, "fourc-configuration.yml"],
            ["bissector", "h k l omega chi phi tth".split(), pytest.raises(KeyError), "", "fourc-configuration.yml"],
        ],
    )
    def test_full_position(mode, keys, context, expected, config_file):
        from ..geom import diffractometer_factory
    
        assert config_file.endswith(".yml")
    
        with context as reason:
            fourc = diffractometer_factory(name="fourc")
>           fourc.operator.restore(HKLPY2_DIR / "tests" / config_file)

hklpy2/tests/test_diffract.py:190: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
hklpy2/ops.py:473: in restore
    self.configuration.restore(file, clear, restore_constraints)
hklpy2/operations/configure.py:107: in restore
    self._fromdict(config, clear, restore_constraints)
hklpy2/operations/configure.py:125: in _fromdict
    self.diffractometer.operator._fromdict(config)
hklpy2/ops.py:131: in _fromdict
    self.constraints._fromdict(config["constraints"])
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = ['-180.0 <= omega <= 180.0', '-180.0 <= chi <= 180.0', '-180.0 <= phi <= 180.0', '-180.0 <= tth <= 180.0']
config = {'chi': {'class': 'LimitsConstraint', 'high_limit': 180.0, 'label': 'chi', 'low_limit': -180.0}, 'phi': {'class': 'LimitsConstraint', 'high_limit': 180.0, 'label': 'phi', 'low_limit': -180.0}, 'theta': {'class': 'LimitsConstraint', 'high_limit': 180.0, 'label': 'theta', 'low_limit': -180.0}, 'ttheta': {'class': 'LimitsConstraint', 'high_limit': 180.0, 'label': 'ttheta', 'low_limit': -180.0}}

    def _fromdict(self, config):
        """Redefine existing constraints from a (configuration) dictionary."""
        for k, v in config.items():
>           self[k]._fromdict(v)
E           KeyError: 'theta'

hklpy2/operations/constraints.py:183: KeyError
```

--- a/hklpy2/operations/configure.py
+++ b/hklpy2/operations/configure.py
@@ -70,7 +70,7 @@ class Configuration:
             y.write("#hklpy2 configuration file\n\n")
             y.write(dump)
 
-    def restore(self, file, clear=True, restore_constraints=True):
+    def restore(self, file, clear=True, restore_constraints=True, solver=None):
         """
         Restore the diffractometer configuration to a YAML file.
 
@@ -123,11 +123,9 @@ class Configuration:
             controls = {}
             oper = self.diffractometer.operator  # alias
             for axis, v in config["constraints"].items():
-                # File might have renamed axes.
-                axis_solver = config["axes"]["axes_xref"][axis]
-                # Local diffractometer might have renamed axes.
-                axis_local = oper.axes_xref_reversed[axis_solver]
-                v["label"] = axis_local
+                axis_canonical = config["axes"]["axes_xref"][axis]
+                axis_local = oper.axes_xref_reversed[axis_canonical]
+                v["label"] = axis_local  # must match
                 controls[axis_local] = v
             config["constraints"] = controls
         else:

--- a/hklpy2/operations/configure.py
+++ b/hklpy2/operations/configure.py
@@ -119,7 +119,18 @@ class Configuration:
             self.diffractometer.operator.reset_constraints()
             self.diffractometer.operator.reset_samples()
 
-        if not restore_constraints:
+        if restore_constraints:
+            controls = {}
+            oper = self.diffractometer.operator  # alias
+            for axis, v in config["constraints"].items():
+                # File might have renamed axes.
+                axis_solver = config["axes"]["axes_xref"][axis]
+                # Local diffractometer might have renamed axes.
+                axis_local = oper.axes_xref_reversed[axis_solver]
+                v["label"] = axis_local
+                controls[axis_local] = v
+            config["constraints"] = controls
+        else:
             config["constraints"] = {}
 
         self.diffractometer.operator._fromdict(config)

--- a/hklpy2/operations/constraints.py
+++ b/hklpy2/operations/constraints.py
@@ -52,13 +52,23 @@ class ConstraintBase(ABC):
         result["class"] = self.__class__.__name__
         return result
 
-    def _fromdict(self, config):
+    def _fromdict(self, config, solver=None):
         """Redefine this constraint from a (configuration) dictionary."""
+        from ..backends import SolverBase
+
         if self.label != config["label"] or self.__class__.__name__ != config["class"]:
             raise ConfigurationError(
                 f"Wrong configuration for {self.__class__.__name__}({self.label!r})."
                 f" Received configuration: {config!r}"
             )
+
+        if isinstance(solver, SolverBase):
+            # Validate with solver.
+            axis = config["label"]
+            if axis not in solver.real_axis_names:
+                raise KeyError(
+                    f"Constraint label {axis=} not in {solver.real_axis_names}."
+                )
         for k in self._fields:
             if k in config:
                 setattr(self, k, config[k])
@@ -177,10 +187,10 @@ class RealAxisConstraints(dict):
         """Return all constraints as a dictionary."""
         return {k: c._asdict() for k, c in self.items()}
 
-    def _fromdict(self, config):
+    def _fromdict(self, config, solver=None):
         """Redefine existing constraints from a (configuration) dictionary."""
         for k, v in config.items():
-            self[k]._fromdict(v)
+            self[k]._fromdict(v, solver=solver)
 
     def valid(self, **reals: Dict[str, NUMERIC]) -> bool:
         """Are all constraints satisfied?"""

--- a/hklpy2/operations/reflection.py
+++ b/hklpy2/operations/reflection.py
@@ -72,13 +72,21 @@ class Reflection:
         pseudo_axis_names: list,
         real_axis_names: list,
         digits: int = 4,
-        solver: object = None,
+        operator: object = None,
     ) -> None:
-        from ..backends import SolverBase
+        from ..ops import Operations
 
-        if isinstance(solver, SolverBase):
-            # What if axes names in wrong sequence?  What if axes renamed?
-            pass  # TODO #23 validate reflection
+        if isinstance(operator, Operations):
+            # What if axes names in wrong sequence?  Required order is assumed.
+            # What if axes renamed?  All reflections must use the same real_axis_names.
+            axes_diffractometer = operator.diffractometer.real_axis_names
+            axes_solver = operator.solver.real_axis_names
+            if real_axis_names not in (axes_diffractometer, axes_solver):
+                raise ReflectionError(
+                    f"{real_axis_names=}"
+                    f" do not match diffractometer ({axes_diffractometer})"
+                    f" or solver ({axes_solver})."
+                )
 
         self.digits = digits
         self.geometry = geometry
@@ -286,9 +294,22 @@ class ReflectionsDict(dict):
         self.prune()
         return {v.name: v._asdict() for v in self.values()}
 
-    def _fromdict(self, config, solver=None):
+    def _fromdict(self, config, operator=None):
         """Add or redefine reflections from a (configuration) dictionary."""
-        for k, refl_config in config.items():
+        from ..ops import Operations
+
+        for refl_config in config.values():
+            if isinstance(operator, Operations):
+                # Remap the names of all the real axes to the current solver.
+                # Real axes MUST be specified in the order specified by the solver.
+                refl_config["reals"] = {
+                    axis: value
+                    for axis, value in zip(
+                        operator.solver.real_axis_names,
+                        refl_config["reals"].values(),
+                    )
+                }
+
             reflection = Reflection(
                 refl_config["name"],
                 refl_config["pseudos"],
@@ -298,7 +319,7 @@ class ReflectionsDict(dict):
                 pseudo_axis_names=list(refl_config["pseudos"]),
                 real_axis_names=list(refl_config["reals"]),
                 digits=refl_config["digits"],  # TODO: Digits are optional?
-                solver=solver,
+                operator=operator,
             )
             self.add(reflection, replace=True)
 

--- a/hklpy2/operations/reflection.py
+++ b/hklpy2/operations/reflection.py
@@ -72,7 +72,14 @@ class Reflection:
         pseudo_axis_names: list,
         real_axis_names: list,
         digits: int = 4,
+        solver: object = None,
     ) -> None:
+        from ..backends import SolverBase
+
+        if isinstance(solver, SolverBase):
+            # What if axes names in wrong sequence?  What if axes renamed?
+            pass  # TODO #23 validate reflection
+
         self.digits = digits
         self.geometry = geometry
         self.name = name
@@ -279,7 +286,7 @@ class ReflectionsDict(dict):
         self.prune()
         return {v.name: v._asdict() for v in self.values()}
 
-    def _fromdict(self, config):
+    def _fromdict(self, config, solver=None):
         """Add or redefine reflections from a (configuration) dictionary."""
         for k, refl_config in config.items():
             reflection = Reflection(
@@ -288,11 +295,10 @@ class ReflectionsDict(dict):
                 refl_config["reals"],
                 wavelength=refl_config["wavelength"],
                 geometry=refl_config["wavelength"],
-                # TODO: #23 What if axes names appear in wrong sequence?
                 pseudo_axis_names=list(refl_config["pseudos"]),
-                # TODO: #23 What if axes renamed?
                 real_axis_names=list(refl_config["reals"]),
-                digits=refl_config["digits"],  # TODO: Optional?
+                digits=refl_config["digits"],  # TODO: Digits are optional?
+                solver=solver,
             )
             self.add(reflection, replace=True)
 

--- a/hklpy2/operations/sample.py
+++ b/hklpy2/operations/sample.py
@@ -77,12 +77,12 @@ class Sample:
             "digits": self.digits,
         }
 
-    def _fromdict(self, config, solver=None):
+    def _fromdict(self, config, operator=None):
         """Redefine sample from a (configuration) dictionary."""
         self.name = config["name"]
         self.digits = config["digits"]
         self.lattice._fromdict(config["lattice"])
-        self.reflections._fromdict(config["reflections"], solver=solver)
+        self.reflections._fromdict(config["reflections"], operator=operator)
         self.reflections.order = config["reflections_order"]
         self.U = config["U"]
         self.UB = config["UB"]

--- a/hklpy2/operations/sample.py
+++ b/hklpy2/operations/sample.py
@@ -77,12 +77,12 @@ class Sample:
             "digits": self.digits,
         }
 
-    def _fromdict(self, config):
+    def _fromdict(self, config, solver=None):
         """Redefine sample from a (configuration) dictionary."""
         self.name = config["name"]
         self.digits = config["digits"]
         self.lattice._fromdict(config["lattice"])
-        self.reflections._fromdict(config["reflections"])
+        self.reflections._fromdict(config["reflections"], solver=solver)
         self.reflections.order = config["reflections_order"]
         self.U = config["U"]
         self.UB = config["UB"]

--- a/hklpy2/ops.py
+++ b/hklpy2/ops.py
@@ -129,7 +129,7 @@ class Operations:
             self.add_sample(key, 1, replace=True)._fromdict(sample)
 
         self.constraints._fromdict(config["constraints"])
-        # TODO: #8 extras & mode
+        # TODO: mode & extras
 
     def add_reflection(
         self,

--- a/hklpy2/ops.py
+++ b/hklpy2/ops.py
@@ -126,10 +126,19 @@ class Operations:
     def _fromdict(self, config):
         """Redefine diffractometer from a (configuration) dictionary."""
         for key, sample in config["samples"].items():
-            self.add_sample(key, 1, replace=True)._fromdict(sample)
+            sample_object = self.add_sample(key, 1, replace=True)
+            sample_object._fromdict(sample, solver=self.solver)
 
-        self.constraints._fromdict(config["constraints"])
-        # TODO: mode & extras
+        for key, constraint in config["constraints"].items():
+            if (
+                constraint["class"] == "LimitsConstraint"
+                and constraint["label"] in config["axes"]["real_axes"]
+            ):
+                # By convention, the 'key' here is the axis name when config was written.
+                axis_canonical = config["axes"]["axes_xref"][key]
+                axis_local = self.axes_xref_reversed[axis_canonical]
+                constraint["label"] = axis_local
+        self.constraints._fromdict(config["constraints"], solver=self.solver)
 
     def add_reflection(
         self,
@@ -470,7 +479,7 @@ class Operations:
 
         Note: Can't name this method "import", it's a reserved Python word.
         """
-        self.configuration.restore(file, clear, restore_constraints)
+        self.configuration.restore(file, clear, restore_constraints, solver=self.solver)
 
     def set_solver(
         self,

--- a/hklpy2/ops.py
+++ b/hklpy2/ops.py
@@ -127,7 +127,7 @@ class Operations:
         """Redefine diffractometer from a (configuration) dictionary."""
         for key, sample in config["samples"].items():
             sample_object = self.add_sample(key, 1, replace=True)
-            sample_object._fromdict(sample, solver=self.solver)
+            sample_object._fromdict(sample, operator=self)
 
         for key, constraint in config["constraints"].items():
             if (

--- a/hklpy2/tests/test_diffract.py
+++ b/hklpy2/tests/test_diffract.py
@@ -181,14 +181,11 @@ def test_diffractometer_wh(capsys):
             None,
             "e4cv_orient.yml",
         ],
-        # FIXME: #23 next test (has renamed axes) should pass with does_not_raise()
-        # ["bissector", "h k l omega chi phi tth".split(), does_not_raise(), None, "fourc-configuration.yml"],
-        # Bypass now with test that validates as-is.
         [
             "bissector",
             "h k l omega chi phi tth".split(),
-            pytest.raises(KeyError),
-            "",
+            does_not_raise(),
+            None,
             "fourc-configuration.yml",
         ],
     ],

